Add Bidirectional Search Implementation

Manthan-R-Sheth opened the pull request:
The PR is aimed at implementing Bidirectional Search for the AIMA4e build.

Manthan-R-Sheth committed:
Add Bidirectional Search Implementation along with unit tests and documentation.

Manthan-R-Sheth commented:
I shall be writing comments in the code and unit tests once the code is good to go.

Manthan-R-Sheth commented:
@ctjoreilly  if you could review this implementation.

ctjoreilly commented:
This overlaps with pull request #242 , so I need to look at both together.

ctjoreilly commented:
Also, could you include a Unit test and fully document your implemtation.

Thanks

Ciaran

Manthan-R-Sheth commented:
@ctjoreilly  ya i will do so and update the PR.
Thanks.

Manthan-R-Sheth commented:
@ctjoreilly i have updated the PR, if you can review and give your feedback. 
I have added unit test, and have extended BasicProblem with BidirectionalProblem to improve its functionality and allowing it to be generic as well.
Thanks.

globalworming commented on core/src/main/java/aima/core/environment/support/ProblemFactory.java:
```diff
@@ -13,10 +13,7 @@
-import aima.core.environment.map2d.InState;
-import aima.core.environment.map2d.Map2DFunctionFactory;
-import aima.core.environment.map2d.SimplifiedRoadMapOfPartOfRomania;
+import aima.core.environment.map2d.*;
```
start imports: unnecessary change and bad practice IMHO
http://stackoverflow.com/a/147461/1359903
I'd revert that

globalworming commented on test/src/test/java/aima/test/unit/search/uninformed/BidirectionalSearchTest.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @Test
+    public void testBidirectionalSearch() {
```
unittest tests 2 things, should be split into "start is goal" test and "from sibiu to fagaras" test

Manthan-R-Sheth replied:
I though to do so, but according to the previous practise followed as seen [here](https://github.com/aimacode/aima-java/blob/AIMA4e/test/src/test/java/aima/test/unit/search/uninformed/BreadthFirstSearchTest.java), I combined the two tests into one.
@globalworming

globalworming commented on core/src/main/java/aima/core/environment/support/ProblemFactory.java:
```diff
@@ -51,6 +48,21 @@
 				Map2DFunctionFactory.getStepCostFunction(simplifidRoadMapOfPartOfRomania));
 	}
 
+	public static Problem<GoAction, InState> getBidirectionalSearch(String initialState, final String goalLocations) {
```
naming: this method does not return a BiDirectionalSearch as the name suggests but a BasicProblem

globalworming reviewed (changes requested):
unittests could test more cases e.g. zerind->sibiu and arad->neamt and arad->sibiu
use concrete imports instead of star imports

Manthan-R-Sheth commented:
@globalworming 
Updated the PR.

globalworming commented on core/src/main/java/aima/core/search/basic/support/BasicProblem.java:
```diff
@@ -2,23 +2,17 @@
-import aima.core.search.api.ResultFunction;
-import aima.core.search.api.ResultsFunction;
-import aima.core.search.api.StepCostFunction;
+import aima.core.search.api.*;
```
Please Check your other changes for Star imports

Manthan-R-Sheth replied:
Fixed the star imports. @globalworming, 
Actually Alt+Enter of IntelliJ does that. I will take care henceforth.

ctjoreilly commented:
I am having problems applying your patch, due to the multiple commits. Did you do any force commits to your pull request? Could I ask you to collapse all your changes into a single squashed pull request?

Manthan-R-Sheth commented:
@ctjoreilly
I will rebase and squash my commits and update the PR.

Manthan-R-Sheth commented:
@ctjoreilly
I have squashed my commits and updated the PR.

ctjoreilly commented on core/src/main/java/aima/core/environment/support/ProblemFactory.java:
```diff
@@ -51,6 +51,21 @@
 	}
 
+	public static Problem<GoAction, InState> getSimpleBidirectionalSearchProblem(String initialState, final String goalLocations) {
+		final SimplifiedRoadMapOfPartOfRomania simplifidRoadMapOfPartOfRomania = new SimplifiedRoadMapOfPartOfRomania();
```
goalLocations -> goalLocation, i.e. should be singular.

ctjoreilly commented on core/src/main/java/aima/core/search/basic/support/BasicProblem.java:
```diff
@@ -102,4 +114,14 @@ public boolean isGoalState(S state) {
+
+	@Override
+	public Problem<A, S> getReverseProblem() {
+		return new BasicProblem<A, S>(finalState, initialState, goalTestPredicate, actionsFn, resultFn, stepCostFn);
```
More efficient to instantiate just once and keep as an attribute (see earlier comments).

ctjoreilly commented on core/src/main/java/aima/core/search/api/BidirectionalProblem.java:
```diff
@@ -0,0 +1,11 @@
+/**
+ * Created by manthan on 2/3/17.
+ */
+
```
Change to '@author manthan'

ctjoreilly replied:
Remove blank line.

ctjoreilly commented on core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:
```diff
@@ -0,0 +1,127 @@
+        if (problem.isGoalState(node.state())) {
+            return solution(node);
+        }
+        Node<A, S> revNode = newRootNode(((BidirectionalProblem<A, S>) problem).getReverseProblem().initialState(), 0);
```
TODO - we need to come up with a way to avoid this dirty cast, suggestions?

ctjoreilly replied:
For now I would suggest adding a SearchForActionsBidirectionallyFunction interface similar to SearchForActionsFunction but that takes a BidirectionalProblem. Update all related code appropriately.

Manthan-R-Sheth replied:
@ctjoreilly I will have to change ```NodeFactory<A,S>``` for making a new child node since now the problem is ```BidirectionalProblem``` and not Problem
In my opinion, ```Node<A, S> newChildNode(Problem<A, S> problem, Node<A, S> parent, A action)``` should besuch that it includes both types of problems.

Manthan-R-Sheth replied:
Can I have ```Node<A, S> newChildNodeForBidirectionalSearch(BidirectionalProblem<A, S> problem, Node<A, S> parent, A action)``` as a function of ```NodeFactory<A,S>``` ?

ctjoreilly commented on core/src/main/java/aima/core/environment/support/ProblemFactory.java:
```diff
@@ -51,6 +51,21 @@
 				Map2DFunctionFactory.getStepCostFunction(simplifidRoadMapOfPartOfRomania));
 	}
 
+	public static Problem<GoAction, InState> getSimpleBidirectionalSearchProblem(String initialState, final String goalLocations) {
```
Should return a BidirectionalProblem.

ctjoreilly commented on core/src/main/java/aima/core/search/basic/support/BasicProblem.java:
```diff
@@ -1,24 +1,26 @@
  */
-public class BasicProblem<A, S> implements Problem<A, S>, NondeterministicProblem<A, S>, OnlineSearchProblem<A, S> {
+public class BasicProblem<A, S> implements Problem<A, S>, NondeterministicProblem<A, S>, OnlineSearchProblem<A, S>, BidirectionalProblem<A, S> {
 	private S initialState;
```
Declare implements BidirecationalProblem after Problem<A, S>, i.e.:

implements Problem<A, S>, Bidirectional....

so that is matches the order these interfaces are introduced in the book.

ctjoreilly commented on core/src/main/java/aima/core/search/basic/support/BasicProblem.java:
```diff
@@ -1,24 +1,26 @@
-public class BasicProblem<A, S> implements Problem<A, S>, NondeterministicProblem<A, S>, OnlineSearchProblem<A, S> {
+public class BasicProblem<A, S> implements Problem<A, S>, NondeterministicProblem<A, S>, OnlineSearchProblem<A, S>, BidirectionalProblem<A, S> {
 	private S initialState;
+	private S finalState;
```
change this to:

private Problem<A, S> reverseProblem and instantiate in the Bidirectional constructor, so that it only needs to be done so once.

ctjoreilly commented on core/src/main/java/aima/core/search/basic/support/BasicProblem.java:
```diff
@@ -73,6 +75,16 @@ public BasicProblem(ActionsFunction<A, S> actionsFn, GoalTestPredicate<S> goalTe
 		this.stepCostFn = stepCostFn;
 	}
 
+	// BidirectionalProblem constructor(Reverse Problem)
```
Move this to below the problem constructors (and ensure you have a blank line between method definitions). Again so that the order matches that of the book in terms of introduction to concepts.

ctjoreilly commented on core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:
```diff
@@ -0,0 +1,127 @@
+ * the bidirectional search algorithm i.e. Problem is reversed and search is performed from both
+ * frontiers.
+ * Created by manthan on 1/3/17.
+ */
```
Change to '@author manthan'

ctjoreilly commented on core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:
```diff
@@ -0,0 +1,127 @@
+                List<A> sol = solution(solutionNode);
+                Collections.reverse(sol);
+                return sol;
+            }
```
Is there a cleaner way to do this without repeating the conditional block twice?

Manthan-R-Sheth replied:
I am not sure about an alternative way for this. This is because alternate evaluations of both queues need to be done for the algorithm.

ctjoreilly commented on core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:
```diff
@@ -0,0 +1,127 @@
+        return null;
+    }
+
+    //Supporting Code
```
Minor tidy up, change to:
//
// Supporting Code

ctjoreilly commented on core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:
```diff
@@ -0,0 +1,127 @@
+
+    public boolean containsState(Queue<Node<A, S>> frontier, S state) {
+        return frontier.stream().anyMatch(frontierNode -> frontierNode.state().equals(state));
+    }
```
May want to consider a more efficient way of doing this. Have a look at BasicFrontierQueue, it will let you do a containment check much faster.

Manthan-R-Sheth replied:
```front``` and ```back``` can be made ```BasicFrontierQueue<A,S>``` . It is good way for improving efficiency.

ctjoreilly commented on test/src/test/java/aima/test/unit/search/uninformed/BidirectionalSearchTest.java:
```diff
@@ -0,0 +1,49 @@
+                searchForActions(ProblemFactory.getSimpleBidirectionalSearchProblem(
+                        SimplifiedRoadMapOfPartOfRomania.ARAD, SimplifiedRoadMapOfPartOfRomania.ARAD)));
+        //Result will contain all intermediate nodes excluding initial and final nodes.
+        Assert.assertEquals(
```
// NOTE: The answer should include the goal as well (i.e. Bucharest) otherwise the list of actions doesn't actually get you there. This should be fixed. I would also consider adding some additional tests.

Manthan-R-Sheth replied:
Since the output is a list of Actions ( ```List<A>``` ), the goal has a null action and so I didn't add it to the result.

ctjoreilly replied:
Thinking about this some more, I think we may need to change the interface on SearchForActionsBidirectionallyFunction to return something other than a List<A>. Don't consider addressing this until we get a basic version merged into the trunk that we agree upon and then we can come back round to see if this idea make sense:

Instead it returns something like this:

 interface BidirectionalActions<A> {

      List<A> fromInitialStatePart()
      List<A> fromGoalStatePart()
      List<A> fromInitialStateToGoalState(); // What I was indicating above I was expecting as  a result
      List<A> fromGoalStateToInitialState(); // essential the reverse of the previous method.
}

Manthan-R-Sheth replied:
Ya i will surely update it when things get finalised.

ctjoreilly reviewed (changes requested):
Hi, 

This looks good as an initial version of a core implementation of the BidirectionalSearch algorithm. Could you please go through the review comments and make the appropriate changes and the re-push (again all squashed together as I apply as a patch).

Thanks

Ciaran

Manthan-R-Sheth committed:
Add interface for bidirectional search,implement changes and add tests

Manthan-R-Sheth commented:
@ctjoreilly if you could review now. 
**Changes made** :- 
1. SearchForActionsBidirectionallyFunction is implemented that takes BidirectionalProblem.
1. newChildNodeForBidirectionalSearch is added to make nodes for bidirectional search.
1. Queues are converted to BasicFrontierQueue for efficient state containment. (can be extended to other classes too).
1. reverseProblem attribute is added in BasicProblem.
1. Unit tests are increased.
1. Minor fixes of @author and new lines.

ctjoreilly commented on core/src/main/java/aima/core/search/api/BidirectionalProblem.java:
```diff
@@ -0,0 +1,20 @@
+        GoalTestPredicate<S>,
+        // double stepCost(S s, A a, S s')
+        StepCostFunction<A, S> {
+    Problem<A,S> getOriginalProblem();
```
Why are you doing the above? Ideally, I'd like to keep this the way it was. I would like to also avoid changing the NodeFactory to take this into account could you explain further?

Manthan-R-Sheth replied:
@ctjoreilly 
I did this because :-

1. newChildNode() function requires a ```Problem```, but we have ```BidirectionalProblem``` as the argument of apply(). Hence to create a node for such case, I modified it accordingly. Could you suggest a better approach.
1. If BidirectionalProblem isn't implemented as above, we won't have functions like ```initialState()```. Thus to make it just like other types of problems like ```Problem```,
``` NondeterministicProblem``` etc I did so.

ctjoreilly replied:
I thought I'd submitted a response as regards this earlier. I do not wish to go with the BidirectionalProblem API in this form or require any changes to the NodeFactory. Could you explain to me the difference between the results from call BidrectionalProblem.initialState() and BidirectionalProblem.getOriginalProblem().initialState()? Similarly, for the other overlapping methods. Basically, I only see BidirectionalProblem needing to be a container for the original and reverse versions of the problem, it itself does not need to be one. Let me know if I'm missing something here.

Thanks

Ciaran

Manthan-R-Sheth replied:
BidirectionalProblem.initialState() and BidirectionalProblem.getOriginalProblem().initialState(), both will have same results. I made BidirectionalProblem.getOriginalProblem() for this purpose only. But I used it as a different problem to keep continuity with the superclasses of BasicProblem (which are all problems in themselves). 
Shall I replace BidirectionalProblem.getOriginalProblem().initialState() ? Your call on this aspect.

@ctjoreilly

ctjoreilly reviewed (changes requested):
See review comment related to BidirectionalProblem interface. Also, this pull request has jumped to 13 files. Could you create a new pull request that only conains your changes (and not those also pulled from the aima4e branch).

Thanks

Manthan-R-Sheth commented:
@ctjoreilly  I have rebased onto my commits and git history is clean.

ctjoreilly commented on core/src/main/java/aima/core/search/basic/support/BasicProblem.java:
```diff
@@ -1,24 +1,26 @@
-public class BasicProblem<A, S> implements Problem<A, S>, NondeterministicProblem<A, S>, OnlineSearchProblem<A, S> {
+public class BasicProblem<A, S> implements Problem<A, S>, BidirectionalProblem<A, S>, NondeterministicProblem<A, S>, OnlineSearchProblem<A, S> {
 	private S initialState;
+	private S finalState;
```
You shouldn't need to track the finalState if you've already instantiated the reverseProblem as I don't see where it will be used otherwise.

Manthan-R-Sheth replied:
Actually it used to initialise the reverse problem , in the constructor.
```		this.reverseProblem = new BasicProblem<>(finalState, actionsFn, resultFn, goalTestPredicate, stepCostFn);```
It acts as the initial state for the reverse problem.
I will remove the global variabe though.

ctjoreilly commented on test/src/test/java/aima/test/unit/search/uninformed/BidirectionalSearchTest.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @Test
+    public void testBidirectionalSearch() {
```
Thinking about this some more, I think we may need to change the interface on SearchForActionsBidirectionallyFunction to return something other than a List<A>. Don't consider addressing this until we get a basic version merged into the trunk that we agree upon and then we can come back round to see if this idea make sense:

Instead it returns something like this:

 interface BidirectionalActions<A> {
      List<A> fromInitialStatePart()
      List<A> fromGoalStatePart()

      List<A> fromInitialStateToGoalState(); // What I was indicating above I was expecting as  a result
      List<A> fromGoalStateToInitialState(); // essential the reverse of the previous method.
}

Manthan-R-Sheth committed:
Remove FinalState variable

Manthan-R-Sheth commented:
@ctjoreilly  Now is it good to go.

ctjoreilly commented:
Hi,

To avoid confusion with the semantic meaning of BidireactionalProblem I think the simplest thing to do is delete it and revert BasicProblem and NodeFactory back to the way they were. Instead modify the apply() method on SearchForActionsBidirectionallyFunction to:

List<A> apply(Problem<A, S> originalProblem, Problem<A, S> reverseProblem);

and the method on ProblemFactory should now return:

Pair<Problem<GoAction, InState>, Problem<GoAction, InState>> getSimpleBidirectionalSearchProblem(...)

This will also help with reducing API bloat. Once done and all tests passing I think this should be good for merging to AIMA4e as the basis for further improvements on BidirectionalSearch.

Thanks

Ciaran

Manthan-R-Sheth committed:
Update the implementation of BidirectionalSearch

ctjoreilly commented on core/src/main/java/aima/core/search/api/SearchForActionsBidirectionallyFunction.java:
```diff
@@ -1,13 +1,15 @@
+public interface SearchForActionsBidirectionallyFunction<A, S> extends Function<Pair<Problem<A, S>, Problem<A, S>>, List<A>> {
     @Override
-    List<A> apply(BidirectionalProblem<A, S> problem);
+    List<A> apply(Pair<Problem<A, S>, Problem<A, S>> problem);
```
This should not take a pair, but broken apart as:

originalProblem, reverseProblem

Using Pair on the ProblemFactory was for convenience as you can't specificy generics in an Array i.e. Problem[].

Manthan-R-Sheth committed:
Convert Pair to different Problems

Manthan-R-Sheth commented:
@ctjoreilly  Reverted back with BasicProblem and NodeFactory.
Updated accordingly.

ctjoreilly commented:
Could you squash/rebase so I can validate the patch before accepting the pull request (almost there :) ).

Thanks

Ciaran

Manthan-R-Sheth commented:
Actually I had rebased locally earlier but forgot to push it on my fork. And I made changes on it locally. I had to merge after that. I cleaned that history by cherrypicking but now I am not able to squash my commits. 
if you could test my PR locally by ```git fetch origin pull/ID/head:BRANCHNAME``` instead of applying patch.
Thanks.

ctjoreilly commented:
Ok, I'll chance a direct merge as I think it should be good (hopefully)

Manthan-R-Sheth commented:
@ctjoreilly 
Thanks.
